bentoml/cli/bento_service.py

```diff
@@ -0,0 +1,536 @@
+                    timeout=timeout,
+                    flags="--enable-microbatch" if enable_microbatch else "",
+                ),
+            )
+
+        if workers is None:
+            workers = get_gunicorn_num_of_workers()
+
+        # Gunicorn only supports POSIX platforms
+        from bentoml.server.gunicorn_server import GunicornBentoServer
+        from bentoml.server.marshal_server import GunicornMarshalServer
+
+        if enable_microbatch:
+            prometheus_lock = multiprocessing.Lock()
+            # avoid load model before gunicorn fork
+            with reserve_free_port() as api_server_port:
+                marshal_server = GunicornMarshalServer(
+                    bundle_path=bento_service_bundle_path,
+                    port=port,
+                    workers=microbatch_workers,
+                    prometheus_lock=prometheus_lock,
+                    outbound_host="localhost",
+                    outbound_port=api_server_port,
+                    outbound_workers=workers,
+                )
+
+                gunicorn_app = GunicornBentoServer(
+                    bento_service_bundle_path,
+                    api_server_port,
+                    workers,
+                    timeout,
+                    prometheus_lock,
+                )
+            marshal_server.async_run()
+            gunicorn_app.run()
+        else:
+            gunicorn_app = GunicornBentoServer(
+                bento_service_bundle_path, port, workers, timeout
+            )
+            gunicorn_app.run()
+
+    @bentoml_cli.command(
+        help="Install shell command completion",
+        short_help="Install shell command completion",
+    )
+    @click.option(
+        '--append/--overwrite',
+        help="Append the completion code to the file",
+        default=None,
+    )
+    @click.argument('shell', required=False, type=shell_types)
+    @click.argument('path', required=False)
+    def install_completion(append, shell, path):
+        if click_completion:
+            # click_completion package is imported
+            shell, path = click_completion.core.install(
+                shell=shell, path=path, append=append
+            )
+            click.echo('%s completion installed in %s' % (shell, path))
+        else:
+            click.echo(
+                "'click_completion' is required for BentoML auto-completion, "
+                "install it with `pip install click_completion`"
+            )
+
+    @bentoml_cli.command(
+        help='Containerizes given Bento into a ready-to-use Docker image.',
+        short_help="Containerizes given Bento into a ready-to-use Docker image",
+    )
+    @click.argument("bento", type=click.STRING)
+    @click.option('-p', '--push', is_flag=True)
+    @click.option(
+        '-t',
+        '--tag',
+        help="Optional image tag. If not specified, Bento will generate one from "
+        "the name of the Bento.",
+        required=False,
```

Comment:
I feel if you could valid this `tag` parameter here with a validation callback here, the rest of the code could be simplified https://click.palletsprojects.com/en/7.x/options/#callbacks-for-validation